source/gameengine/Rasterizer/RAS_CubeMap.cpp

```diff
@@ -0,0 +1,161 @@
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program; if not, write to the Free Software Foundation,
+* Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*
+* Contributor(s): Ulysse Martin, Tristan Porteries.
+*
+* ***** END GPL LICENSE BLOCK *****
+*/
+
+/** \file RAS_CubeMap.cpp
+*  \ingroup bgerast
+*/
+
+#include "RAS_CubeMap.h"
+#include "RAS_Texture.h"
+#include "RAS_IRasterizer.h"
+
+#include "GPU_texture.h"
+#include "GPU_framebuffer.h"
+#include "GPU_draw.h"
+
+#include "DNA_texture_types.h"
+
+#include "glew-mx.h"
+
+static const MT_Matrix4x4 bottomFaceViewMat(
+	-1.0f, 0.0f, 0.0f, 0.0f,
+	0.0f, -1.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, 1.0f, 0.0f,
+	0.0f, 0.0f, 0.0f, 1.0f);
+static const MT_Matrix4x4 topFaceViewMat(
+	1.0f, 0.0f, 0.0f, 0.0f,
+	0.0f, -1.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, -1.0f, 0.0f,
+	0.0f, 0.0f, 0.0f, 1.0f);
+static const MT_Matrix4x4 rightFaceViewMat(
+	1.0f, 0.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, 1.0f, 0.0f,
+	0.0f, -1.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, 0.0f, 1.0f);
+static const MT_Matrix4x4 leftFaceViewMat(
+	1.0f, 0.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, -1.0f, 0.0f,
+	0.0f, 1.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, 0.0f, 1.0f);
+static const MT_Matrix4x4 backFaceViewMat(
+	0.0f, 0.0f, 1.0f, 0.0f,
+	0.0f, -1.0f, 0.0f, 0.0f,
+	1.0f, 0.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, 0.0f, 1.0f);
+static const MT_Matrix4x4 frontFaceViewMat(
+	0.0f, 0.0f, -1.0f, 0.0f,
+	0.0f, -1.0f, 0.0f, 0.0f,
+	-1.0f, 0.0f, 0.0f, 0.0f,
+	0.0f, 0.0f, 0.0f, 1.0f);
+
+MT_Matrix4x4 RAS_CubeMap::facesViewMat[6] = {topFaceViewMat, bottomFaceViewMat, frontFaceViewMat, backFaceViewMat, rightFaceViewMat, leftFaceViewMat};
+MT_Matrix3x3 RAS_CubeMap::camOri[6] = {
+	topFaceViewMat.to3x3(),
+	bottomFaceViewMat.to3x3(),
+	frontFaceViewMat.to3x3(),
+	backFaceViewMat.to3x3(),
+	leftFaceViewMat.to3x3(),
+	rightFaceViewMat.to3x3()
+};
+
+static const GLenum cubeMapTargets[6] = {
+	GL_TEXTURE_CUBE_MAP_POSITIVE_Z_ARB,
+	GL_TEXTURE_CUBE_MAP_NEGATIVE_Z_ARB,
+	GL_TEXTURE_CUBE_MAP_POSITIVE_X_ARB,
+	GL_TEXTURE_CUBE_MAP_NEGATIVE_X_ARB,
+	GL_TEXTURE_CUBE_MAP_POSITIVE_Y_ARB,
+	GL_TEXTURE_CUBE_MAP_NEGATIVE_Y_ARB,
+};
+
+RAS_CubeMap::RAS_CubeMap(RAS_Texture *texture, RAS_IRasterizer *rasty)
+	:m_cubeMapTexture(NULL),
+	m_useMipmap(false),
+	m_texture(texture)
+{
+	EnvMap *env = m_texture->GetMTex()->tex->env;
+
+	m_useMipmap = (env->flag & ENVMAP_MIPMAP) != 0;
+
+	m_cubeMapTexture = m_texture->GetGPUTexture();
+	// Increment reference to make sure the gpu texture will not be freed by someone else.
+	GPU_texture_ref(m_cubeMapTexture);
+
+	if (!m_useMipmap) {
+		// Disable mipmaping.
+		GPU_texture_bind(m_cubeMapTexture, 0);
+		GPU_texture_filter_mode(m_cubeMapTexture, false, false);
```

Comment:
It's not needed to regenerate mipmap each frame when you set the filters min and mag when you set the texture at the beginning (in the contructor). I'll try to find a way to make it as an option.